Source/santad/Logs/EndpointSecurity/Writers/Spool.mm

```diff
@@ -0,0 +1,106 @@
+///
+///     https://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+
+#import "Source/santad/Logs/EndpointSecurity/Writers/Spool.h"
+
+#import "Source/common/SNTLogging.h"
+#include "Source/common/santa_proto_include_wrapper.h"
+
+static const char *kTypeGoogleApisComPrefix = "type.googleapis.com/";
+
+namespace santa::santad::logs::endpoint_security::writers {
+
+std::shared_ptr<Spool> Spool::Create(std::string_view base_dir, size_t max_spool_disk_size,
+                                     size_t max_spool_batch_size, uint64_t flush_timeout_ms) {
+  dispatch_queue_t q = dispatch_queue_create("com.google.santa.daemon.file_base_q",
+                                             DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
+  dispatch_source_t timer_source = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, q);
+  dispatch_source_set_timer(timer_source, dispatch_time(DISPATCH_TIME_NOW, 0),
+                            NSEC_PER_MSEC * flush_timeout_ms, 0);
+
+  auto spool_writer =
+    std::make_shared<Spool>(q, timer_source, base_dir, max_spool_disk_size, max_spool_batch_size);
+
+  spool_writer->BeginFlushTask();
+
+  return spool_writer;
+}
+
+Spool::Spool(dispatch_queue_t q, dispatch_source_t timer_source, std::string_view base_dir,
+             size_t max_spool_disk_size, size_t max_spool_batch_size)
+    : q_(q),
+      timer_source_(timer_source),
+      spool_writer_(base_dir, max_spool_disk_size),
+      log_batch_writer_(&spool_writer_, max_spool_batch_size) {
+  type_url_ = kTypeGoogleApisComPrefix + ::santa::pb::v1::SantaMessage::descriptor()->full_name();
+}
+
+Spool::~Spool() {
+  // Note: `log_batch_writer_` is automatically flushed when destroyed
+  if (!flush_task_started_) {
+    // The timer_source_ must be resumed to ensure it has a proper retain count before being
+    // destroyed. Additionally, it should first be cancelled to ensure the timer isn't ever fired
+    // (see man page for `dispatch_source_cancel(3)`).
+    dispatch_source_cancel(timer_source_);
+    dispatch_resume(timer_source_);
+  }
+}
+
+void Spool::BeginFlushTask() {
+  if (flush_task_started_) {
+    return;
+  }
+
+  std::weak_ptr<Spool> weak_writer = weak_from_this();
+  dispatch_source_set_event_handler(timer_source_, ^{
+    std::shared_ptr<Spool> shared_writer = weak_writer.lock();
+    if (!shared_writer) {
+      return;
+    }
+
+    if (!shared_writer->Flush()) {
+      LOGE(@"Spool writer: periodic flush failed.");
+    }
+  });
+
+  dispatch_resume(timer_source_);
+  flush_task_started_ = true;
+}
+
+bool Spool::Flush() {
+  return log_batch_writer_.Flush().ok();
+}
+
+void Spool::Write(std::vector<uint8_t> &&bytes) {
```

Comment:
It'd be nice if we could have the a variant of `Spool::write` that was able to call a callback e.g. on line 102. 

This would  allow us to avoid sleeps in our code and could be used to update metrics.

Reply:
I see you added this to the constructor great.